src/history_manager.py

```diff
@@ -95,71 +129,53 @@ def rewind_history(self):
 	# Serialized operations ####################################################
 
 	def add_operation(self, operation):
+
 		self._image.set_surface_as_stable_pixbuf()
-		# print('add_operation_to_history')
-		# print(operation['tool_id'])
-		# if 'select' in operation['tool_id']:
-		# 	print(operation['operation_type'])
-		# 	print('-----------------------------------')
+		if self._undo_history[-1].is_full():
+			self.add_state(self._image.main_pixbuf.copy())
+
+		self._undo_history[-1].add_operation(operation)
 		self._is_saved = False
-		self._undo_history.append(operation)
+		self._image.update_title()
+		
 
 	############################################################################
 	# Cached pixbufs ###########################################################
 
 	def set_initial_operation(self, rgba_array, pixbuf, width, height):
+		
 		r = float(rgba_array[0])
 		g = float(rgba_array[1])
 		b = float(rgba_array[2])
 		a = float(rgba_array[3])
-		self.initial_operation = {
+		initial_operation = {
 			'tool_id': None,
 			'pixbuf': pixbuf,
 			'rgba': Gdk.RGBA(red=r, green=g, blue=b, alpha=a),
 			'width': width, 'height': height
 		}
+		self._undo_history.append(State(initial_operation))
 
 	def add_state(self, pixbuf):
 		if pixbuf is None:
 			# Context: an error message
-			raise Exception(_("Attempt to save an invalid state"))
-		self._undo_history.append({
+			raise Exception("Attempt to save an invalid state")
+		
+		new_state = {
 			'tool_id': None,
 			'pixbuf': pixbuf,
 			'width': pixbuf.get_width(),
 			'height': pixbuf.get_height()
-		})
+		}
+
+		self._undo_history.append(State(new_state))
 		self._is_saved = True
 
 	def has_initial_pixbuf(self):
-		return self.initial_operation['pixbuf'] is not None
-
-	def get_last_saved_state(self):
-		index = self._get_last_state_index(False)
-		if index == -1:
-			return self.initial_operation
-		else:
-			return self._undo_history[index]
+		return len(self._undo_history) > 0
 
-	def _get_last_state_index(self, allow_yeeting_states):
-		"""Return the index of the last "state" operation (dict whose 'tool_id'
-		value is None) in the undo-history. If there is no such operation, the
-		returned index is -1 which means the only known state is the
-		self.initial_operation attribute."""
-
-		returned_index = -1
-		nbPixbufs = 0
-		for op in self._undo_history:
-			if op['tool_id'] is None:
-				last_saved_pixbuf_op = op
-				returned_index = self._undo_history.index(op)
-				nbPixbufs += 1
-
-		# TODO if there are too many pixbufs in the history, remove a few ones
```

Comment:
the history was relying mostly on operations rather than states because an implementation such as yours uses a lot of memory. This is extremely visible when editing a huge pic with a system monitor window opened, you'll see "stairs" in the memory graph every 20 operations

the numerous edits, "?", strikethrough text, etc. in issue #200 are mostly related to this dilemma, and my conclusion was that when adding a state, the history manager should try to "remove" a state, or in the case of your implementation, "merge" 2 past states into a single bigger one. This way, the app would still have a CPU performance issue if the user decides to draw a ton of operations and to undo most of them (so a little modal dialog or something like that must be developed to be shown during recomputation) but it would become very rare.

Anyway, i can do the state merging thing myself later, i wrote this explanation just so you know why you should restore this comment line 158, it's still a very valid TODO: with your fix, the history becomes a memory performance issue (instead of a cpu one)

This is also why #200 will still not be closed after merging